pkg/eks/update.go

```diff
@@ -0,0 +1,179 @@
+	if t.skip {
+		return "(skip) " + t.info
+	}
+	return t.info
+}
+func (t *updateClusterConfigTask) Do(errs chan error) error {
+	err := t.call(t.spec)
+	close(errs)
+	return err
+}
+
+// GetCurrentClusterConfigForLogging fetches current cluster logging configuration as two sets - dsiable and enanaled types
+func (c *ClusterProvider) GetCurrentClusterConfigForLogging(cl *api.ClusterMeta) (sets.String, sets.String, error) {
+	enabled := sets.NewString()
+	disabled := sets.NewString()
+
+	cluster, err := c.DescribeControlPlaneMustBeActive(cl)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to retrieve current cluster logging configuration")
+	}
+
+	for _, logTypeGroup := range cluster.Logging.ClusterLogging {
+		for _, logType := range logTypeGroup.Types {
+			if logType == nil {
+				return nil, nil, fmt.Errorf("unexpected response from EKS API - nil string")
+			}
+			if api.IsEnabled(logTypeGroup.Enabled) {
+				enabled.Insert(*logType)
+			}
+			if api.IsDisabled(logTypeGroup.Enabled) {
+				disabled.Insert(*logType)
+			}
+		}
+	}
+	return enabled, disabled, nil
+}
+
+// UpdateClusterConfigForLogging calls UpdateClusterConfig to enable logging
+func (c *ClusterProvider) UpdateClusterConfigForLogging(cfg *api.ClusterConfig) error {
+	all := sets.NewString(api.SupportedCloudWatchClusterLogTypes()...)
+
+	enabled := sets.NewString()
+	if cfg.HasClusterCloudWatchLogging() {
+		enabled.Insert(cfg.CloudWatch.ClusterLogging.EnableTypes...)
+	}
+
+	disabled := all.Difference(enabled)
+
+	input := &awseks.UpdateClusterConfigInput{
+		Name: &cfg.Metadata.Name,
+		Logging: &awseks.Logging{
+			ClusterLogging: []*awseks.LogSetup{
+				{
+					Enabled: api.Enabled(),
+					Types:   aws.StringSlice(enabled.List()),
+				},
+				{
+					Enabled: api.Disabled(),
+					Types:   aws.StringSlice(disabled.List()),
+				},
+			},
+		},
+	}
+
+	output, err := c.Provider.EKS().UpdateClusterConfig(input)
+	if err != nil {
+		return err
+	}
+	if err := c.waitForUpdateToSucceed(cfg.Metadata.Name, output.Update); err != nil {
+		return err
+	}
+
+	describeEnabledTypes := "no types enabled"
+	if len(enabled.List()) > 0 {
+		describeEnabledTypes = fmt.Sprintf("enabled types: %s", strings.Join(enabled.List(), ", "))
+	}
+
+	describeDisabledTypes := "no types disabled"
+	if len(disabled.List()) > 0 {
+		describeDisabledTypes = fmt.Sprintf("disabled types: %s", strings.Join(disabled.List(), ", "))
+	}
+
+	logger.Success("configured CloudWatch logging for cluster %q in %q (%s & %s)",
+		cfg.Metadata.Name, cfg.Metadata.Region, describeEnabledTypes, describeDisabledTypes,
+	)
+	return nil
+}
+
+// UpdateClusterConfigTasks returns all tasks for updating cluster configuration
+func (c *ClusterProvider) UpdateClusterConfigTasks(cfg *api.ClusterConfig) *manager.TaskTree {
+	tasks := &manager.TaskTree{Parallel: false}
+
+	tasks.Append(&updateClusterConfigTask{
+		info: "update CloudWatch logging configurtaion",
+		skip: !cfg.HasClusterCloudWatchLogging(),
```

Comment:
I would really like to avoid the added complexity of the skip option to an already complex system of concurrent tasks. I think this is probably simpler to implement as an `if cfg.HasclustercloudWatchLogging()` inside the actual task.

Adding more complexity to the management of concurrent tasks makes it even harder to reason about the code and hard to debug when there are issues.

Reply:
Maybe we could just rewriting this functions.

Perhaps it could be something like:

```Go
func (c *ClusterProvider) AppendUpdateClusterConfigTasks(cfg *api.ClusterConfig, tasks *manager.TaskTree) {
   newTasks := &manager.TaskTree{Parallel: false}
   if cfg.HasClusterCloudWatchLogging() {
     tasks.Append(&updateClusterConfigTask{
       info: "update CloudWatch logging configurtaion",
     }
   }  else {
     logger.Info("CloudWatch logging will not be enabled for cluster %q in %q", cfg.Metadata.Name, cfg.Metadata.Region)
   if newTasks.Len() == 0 {
     return
   }
   tasks.Append(newTasks)
}
```

Reply:
Keep in mind that we are going to add `UpdateClusterConfigForVPC` soon (for #649), that's why I'm adding new `TaskTree`.